File: 006-zigzag-conversion/zigzag.py
def convert(s, numRows):
    """
    Convert string to zigzag pattern and read line by line.

    Args:
        s: Input string
        numRows: Number of rows in zigzag pattern

    Returns:
        String read line by line from zigzag pattern
    """
    # Edge case: if only 1 row or string length <= numRows, return original string
    if numRows == 1 or len(s) <= numRows:
        return s

    # Create array to store characters for each row
    rows = [''] * numRows

    # Track current row and direction
    current_row = 0
    going_down = True

    # Process each character
    for char in s:
        # Add character to current row
        rows[current_row] += char

        # Move to next row based on direction
        if going_down:
            current_row += 1
        else:
            current_row -= 1

        # Change direction when reaching top or bottom
        if current_row == numRows - 1:  # Reached bottom
            going_down = False
        elif current_row == 0:  # Reached top
            going_down = True

        print(rows)
        print("".join(rows))


    # Combine all rows
    result = ''.join(rows)
    return result


# Test with examples
def test_convert():
    # Example 1
    result1 = convert("PAYPALISHIRING", 3)
    print(f'Example 1: "{result1}"')
    print(f'Expected:  "PAHNAPLSIIGYIR"')
    print(f'Match: {result1 == "PAHNAPLSIIGYIR"}\n')

    # Example 2
    result2 = convert("PAYPALISHIRING", 4)
    print(f'Example 2: "{result2}"')
    print(f'Expected:  "PINALSIGYAHRPI"')
    print(f'Match: {result2 == "PINALSIGYAHRPI"}\n')

    # Example 3
    result3 = convert("A", 1)
    print(f'Example 3: "{result3}"')
    print(f'Expected:  "A"')
    print(f'Match: {result3 == "A"}\n')

    # Additional test
    result4 = convert("ABCD", 2)
    print(f'Additional test: "{result4}"')
    # Pattern would be:
    # A C
    # B D
    # So result should be "ACBD"
    print(f'Expected: "ACBD"')
    print(f'Match: {result4 == "ACBD"}')

# Run tests
test_convert()

File: 006-zigzag-conversion/zigzag.ts
// The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

// P   A   H   N
// A P L S I I G
// Y   I   R
// And then read line by line: "PAHNAPLSIIGYIR"

// Write the code that will take a string and make this conversion given a number of rows:

// string convert(string s, int numRows);

// Example 1:

// Input: s = "PAYPALISHIRING", numRows = 3
// Output: "PAHNAPLSIIGYIR"
// Example 2:

// Input: s = "PAYPALISHIRING", numRows = 4
// Output: "PINALSIGYAHRPI"
// Explanation:
// P     I    N
// A   L S  I G
// Y A   H R
// P     I
// Example 3:

// Input: s = "A", numRows = 1
// Output: "A"

//
// P     I    N
// A   L S  I G
// Y A   H R
// P     I

function convert(s: string, numRows: number): string {
  // Handle edge cases
  if (numRows === 1 || s.length <= numRows) {
    return s;
  }

  // Create array of strings for each row
  const rows: string[] = new Array(numRows).fill("").map(() => "");

  let currentRow = 0;
  let goingDown = true;

  for (let i = 0; i < s.length; i++) {
    rows[currentRow] += s[i];

    if (goingDown) {
      currentRow++;
      if (currentRow === numRows - 1) {
        goingDown = false;
      }
    } else {
      currentRow--;
      if (currentRow === 0) {
        goingDown = true;
      }
    }
  }

  return rows.join("");
}

console.log(convert("PAYPALISHIRING", 2));

// PAHNAPLSIIGYIR
